### ISSUE TITLE
Support ES2020 Optional Chaining Operator

### ISSUE THREAD

**Issue by p-bakker** [mozilla/rhino#937]
Proposal: https://github.com/tc39/proposal-optional-chaining
MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

**Comment by p-bakker** [comment]
See https://github.com/mozilla/rhino/pull/838/files for a previous operator addition

**Comment by ryanphjohnson** [comment]
HI! I'm a JS dev who would love to use Optional Chaining on my current platform which uses Rhino. I'm looking at the attached PR and am attempting to sketch out what the code to make my dream a reality might look like. Before I dive into this I wanted to see if I'm even close enough to attempt such a PR here. Here are my initial high level thoughts on what files would need to change, and what those changes might look like:

src/org/mozilla/javascript/CodeGenerator.java - Needs something like `case Token.OC: `
src/org/mozilla/javascript/Decompiler.java - Needs 
```
            case Token.OC:
                result.append(" ?. ");
                break;
```
src/org/mozilla/javascript/Parser.java - I think this is where the 'real' work needs to be done. Something close to this???
```
switch (tt) {
              case Token.DOT:
              case Token.DOTDOT:
              case Token.OC:
                  lineno = ts.lineno;
                  pn = propertyAccess(tt, pn);
                  pn.setLineno(lineno);
                  break;
...
// Inside this function
private AstNode propertyAccess(int tt, AstNode pn)
...
int token = nextToken();
        switch (token) {
          case Token.THROW:
              // needed for generator.throw();
              
              // Add a check here for if this is Optional Chaining and return 'undefined' here
              // ref = propertyName(-1, 'undefined'); ??? There is probably a function/class for undefined I need to pass in here instead of 'undefined'. The core idea behind optional chaining is we don't throw errors, we return undefined.
              
              saveNameTokenData(ts.tokenBeg, "throw", ts.lineno);               // I assume for Optional Chainging we don't need to call this line ???
              ref = propertyName(-1, memberTypeFlags);
              break;

``` 
src/org/mozilla/javascript/Token.java - Probably needs something here, but I honestly have no clue what. It looks like '?.' would need to be added and assigned a number?

src/org/mozilla/javascript/TokenStream.java - Would need something like this:
```
            case '?': 
                if (matchChar('.')) {
                    return Token.OC;
                }
                return Token.HOOK;
```

src/org/mozilla/javascript/ast/AstNode.java - Needs Something like this:
```
        operatorNames.put(Token.OC, "?.");
```

src/org/mozilla/javascript/resources/Messages.properties - Seems to be the place for error messaging, but I'd have to look more closely as to what errors Optional Chaining could throw


Is this about right? Am I close enough to right that I should go after this, or am I still far enough off base that I should just wait for someone else to take a swing at this? This issue has been open for 2 years so it doesn't feel like this is about to get implemented by someone else.

**Comment by tuchida** [comment]
Nice challenge. You are generally right.

Supplemental information, You may need to modify executor as well. In Rhino, it is `Interpreter` and `BodyCodegen`. These two are switched by `optimizationLevel`. You will need knowledge of the Java bytecode to modify `BodyCodegen`.

I’ll be rooting for your challenge.

**Comment by p-bakker** [comment]
Will be rooting for you as well!

Some newer JS features (like some of the destructuring support) have also been implemented by rewritting the AST, prior to executing it. Maybe Optional Chaining lens itself for that as well.

That way you wouldn't have to mess with the interpreter/codegen. Don't have real experience with doing that, but saw it in the codebase, so you may wanna investigate that route

Optional chaining isn't allowed on the left side of an assignment operator, so somewhere you'll need to check for that.

**Comment by tuchida** [comment]
For now, I have created a flow diagram in a crude manner.
![Screen Shot 2023-06-23 at 15 31 32](https://github.com/mozilla/rhino/assets/201790/ed66a913-1abd-4564-8f0e-3a739ead3134)

**Comment by rbri** [comment]
Have started some work on supporting "||=" and "&&=" having the same in mind - so i think you are on the right track.

@tuchida, @p-bakker maybe we can include some info from this in the developer doc?

**Comment by p-bakker** [comment]
> @tuchida, @p-bakker maybe we can include some info from this in the developer doc?

YES! for now either add to https://github.com/mozilla/rhino/wiki/DRAFT:-Stuff-you'd-want-to-know-about-the-codebase or create a new wiki page, just to record all the info, so when time comes we can move it into the proper place in the documentation

**Comment by rbri** [comment]
maybe Undefined.instance or Undefined.SCRIPTABLE_UNDEFINED are what you are looking for

**Comment by p-bakker** [comment]
Nice progress!

Having a look at the (latest version of the) test262 suit for tests related to the Optional Chaining operator might reveal more of those tricky gotcha's like the lhs assignment

Unfortunately we're not yet on the latest version of test262

**Comment by p-bakker** [comment]
@ryanphjohnson wondering if you made any further progress on this

**Comment by p-bakker** [comment]
Closed by manual merge of #1593

**Comment by ryanphjohnson** [comment]
Sorry for disappearing. I changed jobs somewhat unexpectedly so I put on the back burner and I forgot about it. That's on me. I'm very excited to see this feature implemented though.
